README.rst
Python Client for Cloud Debugger API
====================================

|stable| |pypi| |versions|

`Cloud Debugger API`_: is a feature of Google Cloud Platform that lets you inspect the state of an application, at any code location, without stopping or slowing down the running app. Cloud Debugger makes it easier to view the application state without adding logging statements.

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-debugger-client.svg
   :target: https://pypi.org/project/google-cloud-debugger-client/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-debugger-client.svg
   :target: https://pypi.org/project/google-cloud-debugger-client/
.. _Cloud Debugger API: https://cloud.google.com/debugger/docs
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/clouddebugger/latest
.. _Product Documentation:  https://cloud.google.com/debugger/docs

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Cloud Debugger API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Cloud Debugger API.:  https://cloud.google.com/debugger/docs
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current `active`_ and `maintenance`_ versions of
Python.

Python >= 3.7

.. _active: https://devguide.python.org/devcycle/#in-development-main-branch
.. _maintenance: https://devguide.python.org/devcycle/#maintenance-branches

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an `end-of-life`_
version of Python, we recommend that you update as soon as possible to an actively supported version.

.. _end-of-life: https://devguide.python.org/devcycle/#end-of-life-branches

Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-debugger-client


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-debugger-client

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Cloud Debugger API
   to see other available methods on the client.
-  Read the `Cloud Debugger API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Cloud Debugger API Product documentation:  https://cloud.google.com/debugger/docs
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [1.7.0](https://github.com/googleapis/python-debugger-client/compare/v1.6.1...v1.7.0) (2023-07-06)


### Features

* Enable "rest" transport in Python for services supporting numeric enums ([64dbabb](https://github.com/googleapis/python-debugger-client/commit/64dbabb93040f158fde96467a72b091f70ce7e05))


### Bug Fixes

* Add async context manager return types ([64dbabb](https://github.com/googleapis/python-debugger-client/commit/64dbabb93040f158fde96467a72b091f70ce7e05))


### Documentation

* Fix formatting of request arg in docstring ([64dbabb](https://github.com/googleapis/python-debugger-client/commit/64dbabb93040f158fde96467a72b091f70ce7e05))

## [1.6.1](https://github.com/googleapis/python-debugger-client/compare/v1.6.0...v1.6.1) (2023-01-20)


### Bug Fixes

* Add context manager return types ([3a95f76](https://github.com/googleapis/python-debugger-client/commit/3a95f76295407b60d20c2be35a07611e1ec3ed2d))


### Documentation

* Add documentation for enums ([3a95f76](https://github.com/googleapis/python-debugger-client/commit/3a95f76295407b60d20c2be35a07611e1ec3ed2d))

## [1.6.0](https://github.com/googleapis/python-debugger-client/compare/v1.5.0...v1.6.0) (2023-01-10)


### Features

* Add support for python 3.11 ([#143](https://github.com/googleapis/python-debugger-client/issues/143)) ([3cb3ca6](https://github.com/googleapis/python-debugger-client/commit/3cb3ca68b6f4b8ff4e88d996c7391420498ea35a))

## [1.5.0](https://github.com/googleapis/python-debugger-client/compare/v1.4.3...v1.5.0) (2022-12-15)


### Features

* Add support for `google.cloud.debugger.__version__` ([78ffd89](https://github.com/googleapis/python-debugger-client/commit/78ffd890fba217e03febe8b5ff858515cf43f555))
* Add typing to proto.Message based class attributes ([78ffd89](https://github.com/googleapis/python-debugger-client/commit/78ffd890fba217e03febe8b5ff858515cf43f555))


### Bug Fixes

* Add dict typing for client_options ([78ffd89](https://github.com/googleapis/python-debugger-client/commit/78ffd890fba217e03febe8b5ff858515cf43f555))
* **deps:** Require google-api-core &gt;=1.34.0, >=2.11.0  ([f3dc6f9](https://github.com/googleapis/python-debugger-client/commit/f3dc6f9ede689f48486a9c84c7c233d6508f7fb9))
* Drop usage of pkg_resources ([f3dc6f9](https://github.com/googleapis/python-debugger-client/commit/f3dc6f9ede689f48486a9c84c7c233d6508f7fb9))
* Fix timeout default values ([f3dc6f9](https://github.com/googleapis/python-debugger-client/commit/f3dc6f9ede689f48486a9c84c7c233d6508f7fb9))


### Documentation

* **samples:** Snippetgen handling of repeated enum field ([78ffd89](https://github.com/googleapis/python-debugger-client/commit/78ffd890fba217e03febe8b5ff858515cf43f555))
* **samples:** Snippetgen should call await on the operation coroutine before calling result ([f3dc6f9](https://github.com/googleapis/python-debugger-client/commit/f3dc6f9ede689f48486a9c84c7c233d6508f7fb9))

## [1.4.3](https://github.com/googleapis/python-debugger-client/compare/v1.4.2...v1.4.3) (2022-10-07)


### Bug Fixes

* **deps:** Allow protobuf 3.19.5 ([#133](https://github.com/googleapis/python-debugger-client/issues/133)) ([9f5ea8c](https://github.com/googleapis/python-debugger-client/commit/9f5ea8cafb8f9f5d39e827927b75bc076cba2504))

## [1.4.2](https://github.com/googleapis/python-debugger-client/compare/v1.4.1...v1.4.2) (2022-10-03)


### Bug Fixes

* **deps:** Require protobuf >= 3.20.2 ([#131](https://github.com/googleapis/python-debugger-client/issues/131)) ([dc3dd77](https://github.com/googleapis/python-debugger-client/commit/dc3dd77a7c07ea07c902ea027dcda9581e700c37))

## [1.4.1](https://github.com/googleapis/python-debugger-client/compare/v1.4.0...v1.4.1) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#118](https://github.com/googleapis/python-debugger-client/issues/118)) ([2d56dfa](https://github.com/googleapis/python-debugger-client/commit/2d56dfa0703364b0ed5f28008426cda180ab4403))
* **deps:** require proto-plus >= 1.22.0 ([2d56dfa](https://github.com/googleapis/python-debugger-client/commit/2d56dfa0703364b0ed5f28008426cda180ab4403))

## [1.4.0](https://github.com/googleapis/python-debugger-client/compare/v1.3.2...v1.4.0) (2022-07-15)


### Features

* add audience parameter ([8f2963d](https://github.com/googleapis/python-debugger-client/commit/8f2963d5dcaffec289b4653eafd29aa7915ae16f))


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#110](https://github.com/googleapis/python-debugger-client/issues/110)) ([8f2963d](https://github.com/googleapis/python-debugger-client/commit/8f2963d5dcaffec289b4653eafd29aa7915ae16f))
* require python 3.7+ ([#112](https://github.com/googleapis/python-debugger-client/issues/112)) ([0265d05](https://github.com/googleapis/python-debugger-client/commit/0265d053c9ec5c82e139268388560213e2f32985))

## [1.3.2](https://github.com/googleapis/python-debugger-client/compare/v1.3.1...v1.3.2) (2022-06-03)


### Bug Fixes

* **deps:** require protobuf <4.0.0dev ([#100](https://github.com/googleapis/python-debugger-client/issues/100)) ([2a37072](https://github.com/googleapis/python-debugger-client/commit/2a37072151041db69d2c71fd31a9e5bfa5256218))


### Documentation

* fix changelog header to consistent size ([#101](https://github.com/googleapis/python-debugger-client/issues/101)) ([835df83](https://github.com/googleapis/python-debugger-client/commit/835df83d50beeb0ab452264f2ebb7cc8657b8c0a))

## [1.3.1](https://github.com/googleapis/python-debugger-client/compare/v1.3.0...v1.3.1) (2022-03-05)


### Bug Fixes

* **deps:** require google-api-core>=1.31.5, >=2.3.2 ([#75](https://github.com/googleapis/python-debugger-client/issues/75)) ([26f0b3f](https://github.com/googleapis/python-debugger-client/commit/26f0b3f4f0f0ae325d61bdb69e711dee288d8c93))

## [1.3.0](https://github.com/googleapis/python-debugger-client/compare/v1.2.1...v1.3.0) (2022-02-26)


### Features


owl-bot-staging/v2/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.7",
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.11"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/debugger_v2/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install(
        'mypy',
        'types-requests',
        'types-protobuf'
    )
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
